state/secrets.go

```diff
@@ -2280,7 +2280,16 @@ func (st *State) GrantSecretAccess(uri *secrets.URI, p SecretAccessParams) (err
 		return errors.Errorf("cannot grant access to secret in scope of %q which is not alive", p.Scope)
 	}
 	subjectEntity, subjectCollName, subjectDocID, err := st.findSecretEntity(p.Subject)
-	if err != nil {
+	if p.Subject.Kind() == names.UnitTagKind && errors.Is(err, errors.NotFound) {
+		unitApp, _ := names.UnitApplication(p.Subject.Id())
+		_, err := st.RemoteApplication(unitApp)
+		if err != nil && !errors.Is(err, errors.NotFound) {
+			return errors.Trace(err)
+		}
+		if err == nil {
+			return errors.NotSupportedf("sharing secrets with a unit across a cross model relation")
+		}
```

Comment:
Do we still need to return an error for notfound case?
Probably change below `else if ` to `if`?

Reply:
yes, thank you